src/pymor/reductors/h2.py

```diff
@@ -644,6 +644,201 @@ def reconstruct(self, u):
+        self._store_sigma_b_c(sigma, b, c)
+        assert projection in ('orth', 'Eorth')
+
+        self.logger.info('Starting gap IRKA')
+        self._conv_data = (num_prev + 1) * [None]
+        if conv_crit == 'sigma':
+            self._conv_data[0] = sigma
+        for it in range(maxit):
+            self._set_V_W_reductor(sigma, b, c, projection)
+            rom = self._pg_reductor.reduce()
+            sigma, b, c, gap_rom = self._unstable_rom_to_sigma_b_c(rom)
+            self._store_sigma_b_c(sigma, b, c)
+            self._update_conv_data(sigma, gap_rom, rom, conv_crit)
+            self._compute_conv_crit(rom, conv_crit, it)
+            if self.conv_crit[-1] < tol:
+                break
+
+        return rom
+
+    def _rom0_params_to_sigma_b_c(self, rom0_params):
+        self.logger.info('Generating initial interpolation data')
+        self._check_rom0_params(rom0_params)
+        if isinstance(rom0_params, Integral):
+            sigma, b, c = self._order_to_sigma_b_c(rom0_params)
+        elif isinstance(rom0_params, np.ndarray):
+            sigma = rom0_params
+            _, b, c = self._order_to_sigma_b_c(len(rom0_params))
+        elif isinstance(rom0_params, dict):
+            sigma = rom0_params['sigma']
+            b = rom0_params['b']
+            c = rom0_params['c']
+        else:
+            sigma, b, c, _ = self._unstable_rom_to_sigma_b_c(rom0_params)
+        return sigma, b, c
+
+    def _set_V_W_reductor(self, sigma, b, c, projection):
+        fom = (
+            self.fom.with_(
+                **{op: getattr(self.fom, op).assemble(mu=self.mu)
+                   for op in ['A', 'B', 'C', 'D', 'E']},
+                parameter_space=None,
+            )
+            if self.fom.parametric
+            else self.fom
+        )
+        self.V = tangential_rational_krylov(self.fom.A, self.fom.E, self.fom.B, b, sigma, orth=False)
+        self.W = tangential_rational_krylov(self.fom.A, self.fom.E, self.fom.C, c, sigma, trans=True, orth=False)
+        if projection == 'orth':
+            gram_schmidt(self.V, atol=0, rtol=0, copy=False)
+            gram_schmidt(self.W, atol=0, rtol=0, copy=False)
+        elif projection == 'Eorth':
+            gram_schmidt_biorth(self.V, self.W, product=fom.E, copy=False)
+        self._pg_reductor = LTIPGReductor(fom, self.W, self.V,
+                                          projection == 'Eorth')
+
+    @staticmethod
+    def _unstable_rom_to_sigma_b_c(rom):
+        poles, b, c, gap_rom = GapIRKAReductor._unstable_lti_to_poles_b_c(rom)
+        return -poles, b, c, gap_rom
+
+    @staticmethod
+    def _unstable_lti_to_poles_b_c(rom):
+        A = to_matrix(rom.A, format='dense')
+        B = to_matrix(rom.B, format='dense')
+        C = to_matrix(rom.C, format='dense')
+
+        if isinstance(rom.E, IdentityOperator):
+            P = spla.solve_continuous_are(A.T, C.T, B.dot(B.T), np.eye(len(C)))
```

Comment:
It would be good to add support for other solvers. This looks to me like good motivation to add `algorithms.riccati.solve_ricc_dense` (similar to `algorithms.lyapunov.solve_lyap_dense`).